File: gaming/ref/486_Predict_the_Winner_AlphaBeta.cpp
// https://blog.csdn.net/luke2834/article/details/81807569
class Solution {
public:
    int INF = 2e8+5;
    bool PredictTheWinner(vector<int>& nums) {
        if (!(nums.size() & 1))
            return true;
        int ret = FindMax(nums, 0, nums.size() - 1, -INF, INF, 0);
        return ret >= 0;
    }
    int FindMax(vector<int>& nums, int start, int end, int alpha, int beta, int pre){
        if (start == end){
            return pre + nums[start];
        }
        int ret = FindMin(nums, start + 1, end, alpha, beta, pre + nums[start]);
        alpha = max(alpha, ret);
        if (alpha >= beta)
            return alpha;
        ret = max(ret, FindMin(nums, start, end - 1, alpha, beta, pre + nums[end]));
        return ret;
    }
    int FindMin(vector<int>& nums, int start, int end, int alpha, int beta, int pre){
        if (start == end){
            return pre - nums[start];
        }
        int ret = FindMax(nums, start + 1, end, alpha, beta, pre - nums[start]);
        beta = min(beta, ret);
        if (alpha >= beta)
            return beta;
        ret = min(ret, FindMax(nums, start, end - 1, alpha, beta, pre - nums[end]));
        return ret;
    }
};

File: gaming/486_Predict_the_Winner_AlphaBeta.py
import math
from functools import lru_cache
from typing import List

class Solution:
    def alpha_beta(self, l: int, r: int, curr: int, isMaxPlayer: bool, alpha: int, beta: int) -> int:
        if l == r:
            return curr + self.nums[l] * (1 if isMaxPlayer else -1)

        if isMaxPlayer:
            ret = self.alpha_beta(l + 1, r, curr + self.nums[l], not isMaxPlayer, alpha, beta)
            alpha = max(alpha, ret)
            if alpha >= beta:
                return alpha
            ret = max(ret, self.alpha_beta(l, r - 1, curr + self.nums[r], not isMaxPlayer, alpha, beta))
            return ret
        else:
            ret = self.alpha_beta(l + 1, r, curr - self.nums[l], not isMaxPlayer, alpha, beta)
            beta = min(beta, ret)
            if alpha >= beta:
                return beta
            ret = min(ret, self.alpha_beta(l, r - 1, curr - self.nums[r], not isMaxPlayer, alpha, beta))
            return ret

    def PredictTheWinner(self, nums: List[int]) -> bool:
        self.nums = nums
        v = self.alpha_beta(0, len(nums) - 1, 0, True, -math.inf, math.inf)
        return v >= 0



if __name__ == "__main__":
    # nums = [1, 5, 2]
    nums = [949829,1462205,1862548,20538,8366111,5424892,7189089,9,5301221,5462245,0,2,4,9401420,4937008,1,9,7,4146539]
    s = Solution()
    print(s.PredictTheWinner(nums))

